### ISSUE TITLE
Documentation library messes up the definition of \_

### ISSUE THREAD

**Issue by TheTeXnician** [T-F-S/tcolorbox#112]
The `\tcb@doc@index@doc` command 

https://github.com/T-F-S/tcolorbox/blob/55d9deed444eef58c65a88b893500255f44ac7ba/tex/latex/tcolorbox/tcbdocumentation.code.tex#L883-L901

has multiple issues. First of all, it messes up the definition of `\_` which makes it unusable in regular documents making use of this.

@Skillmon analyzed the problem in https://chat.stackexchange.com/transcript/message/55558281#55558281:

> The issue is that the `\ExplSyntaxOn` can't do its thing inside the macro definition as the contents then are already tokenized, hence `\ExplSyntaxOn` and `\ExplSyntaxOff` have to be put around the macro definition, not inside it, or else `\def\__tcobox_index_command:` and `\def\__tcobox_index_environment:` don't redefine those macros but the macro `\_` (as that's then not of category code letter) and this will then have the required parameter text `_tcobox_index_command:`.

A separate issue is that `SpecialMainEnvIndex` probably misses a backslash in front.

**Comment by muzimuzhi** [comment]
A minimal change could be (noted the four `%` used in definition of `\tcb@doc@index@doc` can be safely removed as well):

```diff
diff --git a/tex/latex/tcolorbox/tcbdocumentation.code.tex b/tex/latex/tcolorbox/tcbdocumentation.code.tex
index 819197e..a416a84 100644
--- a/tex/latex/tcolorbox/tcbdocumentation.code.tex
+++ b/tex/latex/tcolorbox/tcbdocumentation.code.tex
@@ -880,8 +880,8 @@
 
 \let\tcb@doc@index@pgf=\tcb@doc@index@pgfsection%
 
+\ExplSyntaxOn
 \def\tcb@doc@index@doc{%
-  \ExplSyntaxOn
   \def \__tcobox_index_command:
   {
     \iftcb@doc@toindex
@@ -891,14 +891,14 @@
   \def \__tcobox_index_environment:
   {
     \iftcb@doc@toindex
-      SpecialMainEnvIndex{\kvtcb@doc@sortindex\idx@actual\tcbIndexPrintEnvC{\kvtcb@doc@index}}
+      \SpecialMainEnvIndex{\kvtcb@doc@sortindex\idx@actual\tcbIndexPrintEnvC{\kvtcb@doc@index}}
     \fi
   }
-  \ExplSyntaxOff
-  \tcbset{index german settings}%
+  \tcbset{index~ german~ settings}%
   \EnableCrossrefs%
   \PageIndex%
 }
+\ExplSyntaxOff
 
 \def\tcb@doc@index@off{}%
```

**Comment by T-F-S** [comment]
This bug was introduced with the recent expl3 adaptions of the `documentation` library.
I will correct it for the next version. Thank you for reporting.

**Comment by T-F-S** [comment]
Resolved by version `4.32 (2020-09-17)`.
